File: facedetect.py
# -*- coding:utf-8 -*-
import numpy
import cv2
import sys
import os
import os.path
import re
from progressbar import ProgressBar
import time


def input_arg(argvs, argc):
    if (argc != 3):   # 引数が足りない場合は、その旨を表示
        print 'Usage: # python %s srcdirectory outputdirectory' % argvs[0]
        quit()        # プログラムの終了

    print 'Input directoryname = %s' % argvs[1]
    print 'Output directoryname = %s' % argvs[2]
    # 引数でとったディレクトリの文字列をリターン
    return argvs


def input_filename(directory_path):
    # コマンド引数でとったディレクトリに含まれているファイル（文字列を取得）
    # サブディレクトリは想定してない
    files = os.listdir(directory_path)
    file_array = []

    for file in files:
        # 引数でとったディレクトリの文字列とそのディレクトリに入っているファイル名を結合
        file_directory = directory_path + "/" + file
        # 再びリストに格納
        file_array.append(file_directory)

    return file_array


def fild_all_files(directory):
    for root, dirs, files in os.walk(directory):
        yield root
        for file in files:
            yield os.path.join(root, file)


def filenamelist(src_directory):
    # サブディレクトリを想定
    file_array = []
    for file in fild_all_files(src_directory):
        # .pngか.jpgのときリストに格納
        if file.endswith(".jpg") or file.endswith(".png"):
            file_array.append(file)

    return file_array

def create_directory(output_directory):
    if os.path.isdir(output_directory) == 0:
        print "Not exist \"%s\" folder. So create it." % output_directory
        os.makedirs(output_directory)
    else:
        print "Exist \"%s\" folder." % output_directory


#イメージのパスと名前付けに使う数字のはじめの数字
def facedetect(image_path, output_directory, num):
    #ファイル読み込み
    image = cv2.imread(image_path)
    #グレースケール変換
    image_gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    #カスケード分類器の特徴量を取得する
    cascade = cv2.CascadeClassifier(cascade_path)
    #物体認識（顔認識）の実行
    facerect = cascade.detectMultiScale(image_gray, scaleFactor=1.1, minNeighbors=1, minSize=(50, 50))
    if len(facerect) <= 0:
        return 0

    for rect in facerect:
        # print rect
        x = rect[0]
        y = rect[1]
        w = rect[2]
        h = rect[3]
        # 正規表現でファイルネームだけに
        # http://docs.python.jp/2/library/re.html
        image_path = re.sub('^.*/', "", image_path)
        # 拡張子の削除
        image_path = image_path.rstrip(".jpg")
        image_path = image_path.rstrip(".png")
        # ファイルネームの決定
        img_name = "./" + output_directory + "/" + image_path + "_" + str(num) + '.jpg'
        # 出力窓を調整
        cut_img = image[y-h*0.2:y+h*1.2, x-w*0.2:x+h*1.2]
        cv2.imwrite(img_name, cut_img,(100,100))
        num += 1
    return num


if __name__ == "__main__":
    cascade_path = '/usr/local/share/OpenCV/haarcascades/haarcascade_frontalface_alt.xml'

    argvs = sys.argv   # コマンドライン引数を格納したリストの取得
    argc = len(argvs)  # 引数の個数

    directory = input_arg(argvs, argc)
    # まず元画像があるディレクトリを持ってくる
    image_path = directory[1]
    # file_namesにディレクトリ込みのファイルネームを格納
    file_names = filenamelist(image_path)
    #引数にとった出力先ディレクトリの存在確認 & 作成
    create_directory(directory[2])

    i = 0

    if len(file_names) == 0:
        print "Not exist jpg or png file !"
    else :
        p = ProgressBar(len(file_names))
        print "File num = ",len(file_names)

    # 元画像のファイルの数だけ顔認識を続ける
    for file_name in file_names:
        # print file_name
        try:
            facedetect(file_name, directory[2], 1)
        except:
            print "Oops! Illegal image... Skip!!"
        p.update(i+1)
        i += 1

    print "Complete !"

File: facedetect.cpp
#include "opencv2/objdetect/objdetect.hpp"
#include "opencv2/highgui/highgui.hpp"
#include "opencv2/imgproc/imgproc.hpp"

#include <cctype>
#include <iostream>
#include <iterator>
#include <stdio.h>
#include <boost/filesystem.hpp> // ファイル一覧を取得
#include <boost/foreach.hpp>
#include <sys/stat.h> // ディレクトリ・ファイル名の確認
#include <sys/types.h> // ディレクトリ作成

using namespace std;
using namespace cv;
using namespace boost;

namespace fs = boost::filesystem;

static void help()
{
  cout <<   "./facedetect --input=\"directory\" --output=\"directory\"\n\n"
            "During execution:\n\tHit any key to quit.\n"
            "\tUsing OpenCV version " << CV_VERSION << "\n" << endl;
}

void detectAndSave(CascadeClassifier &cascade, string &srcfilename, string &outputpath);
string ExtractFileName(const string &path, bool without_extension);
bool isInImage(cv::Mat &img, cv::Rect &rect);
cv::Rect fixRectSize(cv::Mat &img, cv::Rect &rect);

string cascadeName = "/usr/local/share/OpenCV/haarcascades/haarcascade_frontalface_alt.xml";

class Argument
{
public:
  Argument(string name)
	:name_(name), prefix_name_(name), prefix_name_len_(name.length()), flag_(false){}
  ~Argument(){}
  int prefix_compare(char* argv){
	int result = prefix_name_.compare(0, prefix_name_len_, argv, prefix_name_len_);
	if(result == 0) { flag_ = true; }
	return result
  }
  void rm_prefix(){
	name_.assign(prefix_name_.c_str() + prefix_name_len_);
  }
private:
  string prefix_name_;
  stirng prefix_name_len_;
  stirng name_;
  bool flag_;
};


class ArgumentParser
{
public:
  ArgumentParser(std::vector<string> options):
	inputDir_(options[0]), outputDir_(options[1]){}
  ~ArgumentParser(){}
  perse(int argc, const char** argv){
	for( int i = 1; i < argc; i++ ){
	  if( inputDir_.prefix_compare( argv[i] ) == 0){
		inputDir_.rm_prefix();
	  }
	  else if( outputDir_.prefix_compare( argv[i] ) == 0){
		outputDir_.rm_prefix();
	  }else{
		cerr << "WARNING: Unknown option %s" << argv[i] << endl;
	  }
    }
  }
private:
  Argument inputDir_;
  Argument outputDir_;
};

int main( int argc, const char** argv )
{
  Mat frame, frameCopy, image;
  string inputDir = "--input=";
  size_t inputDirLen = inputDir.length();
  bool isInputDir = false;
  string outputDir = "--output=";
  size_t outputDirLen = outputDir.length();
  bool isOutputDir = false;
  CascadeClassifier cascade;
  string inputName;

  help();

  // 引数処理
  for( int i = 1; i < argc; i++ )
    {
	  cout << "Processing " << i << " " <<  argv[i] << endl;
	  if( inputDir.compare( 0, inputDirLen, argv[i], inputDirLen) == 0)
		{
		  inputDir.assign( argv[i] + inputDirLen);
		  cout << "input directory = " << inputDir << endl;
		  isInputDir = true;
		}
	  else if( outputDir.compare( 0, outputDirLen, argv[i], outputDirLen) == 0)
		{
		  outputDir.assign( argv[i] + outputDirLen);
		  cout << "output directory = " << outputDir << endl;
		  isOutputDir = true;
		}
	  else if( argv[i][0] == '-' )
        {
		  cerr << "WARNING: Unknown option %s" << argv[i] << endl;
        }
	  else
		inputName.assign( argv[i] );
    }

  // 入力画像ディレクトリの確認
  if(! isInputDir)
	{
	  cerr << "No input directory... " << endl;
	  help();
	  return -1;
	}
  else
	{
	  struct stat st;
	  if(stat(inputDir.c_str(), &st) != 0)
		{
		  cerr << "No such a directory " << inputDir << endl;
		  return -1;
		}
	}

  // 出力画像ディレクトリの確認
  if(! isOutputDir)
	{
	  cerr << "No output directory" << endl;
	  cout << "will make a directory \"output\"" << endl;
	  mkdir("output", 0755); // "outputディレクトリを作成"
	  outputDir.assign("output" + outputDirLen);
	}
  else
	{
	  struct stat st;
	  if(stat(outputDir.c_str(), &st) != 0)
		{
		  cerr << "No such a directory " << outputDir << endl;
		  cout << "will make a directory "  << outputDir << endl;
		  mkdir(outputDir.c_str(), 0755);
		}
	}

  if( !cascade.load( cascadeName ) )
    {
	  cerr << "ERROR: Could not load classifier cascade" << endl;
	  cout << "Please Check a cascade file is exist at " << 
		"\"" << cascadeName << "\"" << endl;
	  return -1;
    }

  //inputディレクトリのファイル一覧を取得する
  std::vector<std::string> file_list;
  const fs::path path(inputDir);
  BOOST_FOREACH(const fs::path& p, std::make_pair(fs::directory_iterator(path),
												  fs::directory_iterator())) {
	if (!fs::is_directory(p))
	  file_list.push_back(p.string());
	std::cout << p.filename() << std::endl;
  }
  
  for(vector<string>::iterator iter = file_list.begin(); iter != file_list.end(); iter++){
	detectAndSave(cascade, *iter, outputDir);	
  }


  return 0;

}

// 顔検出と保存
void detectAndSave(CascadeClassifier &cascade, string &srcfilename, string &outputpath)
{
  Mat inputImage = imread(srcfilename, 1);
  Mat grayImage;
  cvtColor(inputImage, grayImage,  CV_RGB2GRAY);
  Mat eqalizedImage = inputImage;
  // equalizeHist( grayImage, eqalizedImage );
  vector<Rect> faces;

  cascade.detectMultiScale( eqalizedImage, faces,
							1.1, 1, 0, Size(1, 1) );
  // 切り出して保存
  int num = 1;
  for (vector<Rect>::iterator iter = faces.begin(); iter != faces.end(); iter ++) {
	string saveImgname = ExtractFileName(srcfilename, true);
	string saveImgpath = outputpath + "/" + saveImgname + "_" + to_string(num) + ".jpg";
	cout << saveImgpath << endl;
	num++;
	Rect newRect = Rect(iter->x - (iter->width)*0.3,
						iter->y - (iter->height)*0.3,
						(iter->width)*1.6,
						(iter->height)*1.6);
   
	cout << *iter << endl;
	cout << newRect << endl;
	Rect fixedRect = fixRectSize(inputImage, newRect);
	cout << fixedRect << endl;cout << endl;
	imwrite(saveImgpath, inputImage(fixedRect));
  }
}

// パスからファイル名を抽出（拡張子を除くフラグ付き）
// http://www.slis.tsukuba.ac.jp/~fujisawa.makoto.fu/cgi-bin/wiki/index.php?string%A4%CB%A4%E8%A4%EB%CA%B8%BB%FA%CE%F3%BD%E8%CD%FD#w175146e
string ExtractFileName(const string &path, bool without_extension = true)
{
  string fn;
  string::size_type fpos;
  if((fpos = path.find_last_of("/")) != string::npos){
	fn = path.substr(fpos+1);
  }
  else if((fpos = path.find_last_of("\\")) != string::npos){
	fn = path.substr(fpos+1);
  }
  else{
	fn = path;
  }
 
  if(without_extension && (fpos = fn.find_last_of(".")) != string::npos){
	fn = fn.substr(0, fpos);
  }
 
  return fn;
}

// Rectの範囲が画像内かどうかチェック
bool isInImage(cv::Mat &img, cv::Rect &rect)
{
  if(rect.x < 0) return false;
  if(rect.y < 0) return false;
  if(rect.x + rect.width > img.cols) return false;
  if(rect.y + rect.height > img.rows) return false;
  return true;
}

// Rectの範囲が画像内になければ修正する
cv::Rect fixRectSize(cv::Mat &img, cv::Rect &rect)
{
  if(isInImage(img, rect)) return rect;
  cv::Rect fixedRect = rect;
  if(rect.x < 0) fixedRect.x = 0;
  if(rect.y < 0) fixedRect.y = 0;
  if(fixedRect.x + rect.width > img.cols) fixedRect.width = img.cols - fixedRect.x;
  if(fixedRect.y + rect.height > img.rows) fixedRect.height = img.rows - fixedRect.y;
  return fixedRect;
}
